packages/swapper/src/swappers/ButterSwap/swapperApi/getTradeQuote.ts

```diff
@@ -0,0 +1,232 @@
+      }),
+    )
+  }
+
+  // Map gasFee.amount to networkFeeCryptoBaseUnit using fee asset precision
+  const networkFeeCryptoBaseUnit = bnOrZero(route.gasFee?.amount).gt(0)
+    ? toBaseUnit(route.gasFee.amount, feeAsset.precision)
+    : '0'
+
+  const nativeFeeBaseUnit = bnOrZero(route.swapFee.nativeFee).gt(0)
+    ? toBaseUnit(route.swapFee.nativeFee, feeAsset.precision)
+    : '0'
+  const tokenFeeBaseUnit = bnOrZero(route.swapFee.tokenFee).gt(0)
+    ? toBaseUnit(route.swapFee.tokenFee, sellAsset.precision)
+    : '0'
+
+  let protocolFees: Record<string, ProtocolFee> | undefined = undefined
+  const hasNativeFee = bnOrZero(nativeFeeBaseUnit).gt(0)
+  const hasTokenFee = bnOrZero(tokenFeeBaseUnit).gt(0)
+
+  if (hasNativeFee && hasTokenFee && feeAssetId === sellAsset.assetId) {
+    // If both fees are for the same asset, sum them
+    protocolFees = {
+      [feeAssetId]: {
+        amountCryptoBaseUnit: bnOrZero(nativeFeeBaseUnit).plus(tokenFeeBaseUnit).toString(),
+        requiresBalance: true,
+        asset: feeAsset,
+      },
+    }
+  } else {
+    protocolFees = {}
+    if (hasNativeFee) {
+      protocolFees[feeAssetId] = {
+        amountCryptoBaseUnit: nativeFeeBaseUnit,
+        requiresBalance: true,
+        asset: feeAsset,
+      }
+    }
+    if (hasTokenFee) {
+      protocolFees[sellAsset.assetId] = {
+        amountCryptoBaseUnit: tokenFeeBaseUnit,
+        requiresBalance: true,
+        asset: sellAsset,
+      }
+    }
+    if (Object.keys(protocolFees).length === 0) protocolFees = undefined
+  }
+
+  // Calculate rate as lastHop.totalAmountOut / srcChain.totalAmountIn (in base units)
+  const inputAmount = bnOrZero(route.srcChain.totalAmountIn)
+  const outputAmount =
+    route.dstChain?.totalAmountOut ??
+    route.bridgeChain?.totalAmountOut ??
+    route.srcChain.totalAmountOut
+  const rate = inputAmount.gt(0) ? bnOrZero(outputAmount).div(inputAmount).toString() : '0'
+
+  const step = {
+    buyAmountBeforeFeesCryptoBaseUnit: toBaseUnit(outputAmount, buyAsset.precision),
+    buyAmountAfterFeesCryptoBaseUnit: toBaseUnit(outputAmount, buyAsset.precision),
+    sellAmountIncludingProtocolFeesCryptoBaseUnit: amount,
+    feeData: {
+      networkFeeCryptoBaseUnit,
+      protocolFees: undefined,
+    },
+    rate,
+    source: SwapperName.ButterSwap,
+    buyAsset,
+    sellAsset,
+    accountNumber,
+    allowanceContract: route.contract ?? '0x0',
+    estimatedExecutionTimeMs: route.timeEstimated * 1000,
+    butterSwapTransactionMetadata: {
+      to: buildTx.to,
+      data: buildTx.data,
+      value: buildTx.value,
+      chainId: buildTx.chainId,
+      method: buildTx.method,
+      args: buildTx.args?.map(arg => ({
+        type: arg.type,
+        value: Object.prototype.hasOwnProperty.call(arg, 'value') ? arg.value : null,
+      })),
+    },
+  }
+
+  const tradeQuote: TradeQuote = {
+    id: route.hash,
+    rate,
+    receiveAddress,
+    affiliateBps: DEFAULT_BUTTERSWAP_AFFILIATE_BPS.toString(),
```

Comment:
note for the final version of this, should ensure we pass affiliate and fee rate to butter
<img width="801" alt="image" src="https://github.com/user-attachments/assets/a66c19ec-b406-431f-b05c-bc1539ac6234" />

Reply:
yeah - i think someone else can PR this in later when affiliate address is all setup and ready? @0xApotheosis is doing the back and forth with them, some issues re walletconnect/safe, suspect this code will be mergable prior to that being sorted.

Reply:
Yeah I'm still working with Butter to get an affiliate code set up. 
That said, for this PR maybe just use `'butter'` and we'll update it when we get our own.

Reply:
@0xApotheosis done

Reply:
Sounds sane to me sers.